Shrine Schedule: Addresses Support Ticket: Scheduled Shrine Boosts (#626)

milanig-creator opened the pull request:
# Pull Request

**Please fill: what was implemented, why, possibly include screenshots, are any of the changes breaking, etc.**
## Summary

This PR implements **scheduled Shrine Boosts**, addressing Support Ticket **#626**.

Village leaders can now schedule shrine boosts to activate and expire automatically at defined times, removing the need for manual activation and ensuring consistency across restarts and deployments.

---

## What Was Implemented

- Added a new `ShrineBoostSchedule` table to persist scheduled boosts
- Implemented server-side scheduling logic to:
  - Activate boosts when their start time is reached
  - Expire boosts cleanly when their duration ends
- Integrated the scheduler with the existing Shrine system
- Updated Shrine UI to display and manage scheduled boosts
- Added cron configuration to ensure scheduled boosts are processed reliably in production

---

## Why This Change

Previously, shrine boosts required manual activation and could not be reliably scheduled in advance.  
This caused operational friction for village leadership and inconsistencies during downtime or missed activations.

This change provides:
- Predictable boost timing
- Reduced manual intervention
- Better long-term scalability for shrine mechanics

---

## Testing

- Tested locally and in dev
- Walked through scheduling, activation, and expiration flows
- Confirmed boosts activate and expire as expected
- Verified UI reflects scheduled and active states correctly

---

## Breaking Changes

- None expected

---

## Notes for Reviewers

- Scheduler logic is defensive and idempotent
- UI changes are null-safe if no schedules exist
- Open to follow-up refinements if preferred patterns differ

## License

By making this pull request, I confirm that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the Studie-Tech ApS organization has the copyright to use and modify my contribution for perpetuity.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Shrine boost scheduling: calendar/time picker, per-boost scheduling UI, and APIs to list/schedule/cancel boosts.
  * New Calendar component for date selection.

* **Improvements**
  * Minute-by-minute maintenance tick for more responsive boost activation and cleanup.
  * Shrine Hall UI: activate now or schedule boosts, manage scheduled boosts, improved active-boost display.
  * Stronger concurrency and token safeguards across shrine actions.

* **Chores**
  * New time utilities and a minute-based timer utility.
  * Increased maintenance cron frequency.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces persistence for scheduled shrine boosts and refines related constants/guidelines.
> 
> - Creates `ShrineBoostSchedule` table (with indexes on `villageId`, `boostType`, `startAt`, `endAt`) to store scheduled boosts
> - Adds `MAX_BOOSTS_PER_SHRINE = 50` in `app/drizzle/constants.ts`
> - Updates `CLAUDE.md` DB patterns to emphasize minimizing database roundtrips (prefer parallel queries with `Promise.all()`)
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 609ce63fadaa35c4e7bf92a586f7890d7fb74552. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

milanig-creator commented on app/src/app/api/shrine-boost-scheduler/route.ts:
```diff
@@ -0,0 +1,304 @@
+import { shrineBoostSchedule, village } from "@/drizzle/schema";
+import { and, lte, gt, eq, inArray } from "drizzle-orm";
+import { cookies, headers } from "next/headers";
+import { lockWithMinuteTimer } from "@/libs/gamesettings";
```
Confirmed.

lockWithMinuteTimer has been added to libs/gamesettings.ts in commit c1cfe60, and the scheduler now imports and uses it correctly.

Build and typecheck verified locally.

Phrosfire commented on app/src/app/api/shrine-boost-scheduler/route.ts:
```diff
@@ -0,0 +1,298 @@
+  if (xSecret && xSecret === CRON_SECRET) return true;
+
+  return false;
+}
```
We don't need to use a cron secret for authorization.  This along with the other references here can be removed and it'll work perfectly fine :)

milanig-creator replied:
Thank you, I removed this

Phrosfire commented:
The front end works great for scheduling boosts as well as cancelling them.  I did notice though that when a boost becomes active, it's not being removed from the scheduler.  To avoid clutter and prevent people from cancelling already active/completed boosts for a refund (Won't actually stop an already running boost), we should hide scheduled boosts with a start time earlier than the current time as well as put a safeguard on the delete function so that you can't delete already past/current scheduled boosts.

milanig-creator commented:
> The front end works great for scheduling boosts as well as cancelling them. I did notice though that when a boost becomes active, it's not being removed from the scheduler. To avoid clutter and prevent people from cancelling already active/completed boosts for a refund (Won't actually stop an already running boost), we should hide scheduled boosts with a start time earlier than the current time as well as put a safeguard on the delete function so that you can't delete already past/current scheduled boosts.

I have addressed this and tested in my latest commit. The past schedules are removed from view and added a two-layer safeguard on cancel:

an early check that rejects cancellation if startAt <= now

and a DB-level guard so the delete only succeeds when startAt > now

MathiasGruber commented on app/vercel.json:
```diff
@@ -79,6 +79,10 @@
+    },
+    {
+      "path": "/api/shrine-boost-scheduler",
+      "schedule": "* * * * *"
```
Am I missing something? still seems to be in the shrine-boost-scheduler route?

milanig-creator replied:
This was my fault-- not fully understanding, I'm sorry! I removed the dedicated cron entry from vercel.json and moved the scheduling trigger into the existing shrine maintenance cron.

Its still saying I have this in the file and I cannot find it at all at a second glance. Will look at this in a second

MathiasGruber reviewed (changes requested):
I feel like the whole time-handling is potentially over-engineered. This is a review by opus on the matter, which reflects my own thinking reading through this:

<img width="796" height="1007" alt="image" src="https://github.com/user-attachments/assets/846c573c-cdbd-4347-89da-00554cfeeafd" />

milanig-creator commented on app/drizzle/migrations/0320_cold_mariko_yashida.sql:
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE `ShrineBoostSchedule` (
+	`id` varchar(191) NOT NULL,
```
This is now consolidated into a single commit. 

I believe I addressed what you asked for regarding the migrations, though this part I don't entirely understand. 

I am only seeing 0320_cold_mariko_yashida.sql. To my knowledge, I had initially deleted all migrations from main if I understand my mistake correctly. Im pretty sure this is not what you were asking for, and I reverted this.
